codec/address/bech32_codec.go

```diff
@@ -2,30 +2,79 @@ package address
 
 import (
 	"errors"
+	"fmt"
 	"strings"
+	"sync"
+
+	"github.com/hashicorp/golang-lru/simplelru"
 
 	"cosmossdk.io/core/address"
 	errorsmod "cosmossdk.io/errors"
 
+	"github.com/cosmos/cosmos-sdk/internal/conv"
 	sdkAddress "github.com/cosmos/cosmos-sdk/types/address"
 	"github.com/cosmos/cosmos-sdk/types/bech32"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 )
 
+var errEmptyAddress = errors.New("empty address string is not allowed")
+
+type Option func(*Options)
+
+type Options struct {
```

Comment:
nit: I feel like we''re mixing paradigms here isn't either an `Options` struct with exported fields or composable functions on exported fields in `cachedBech32Codec` a bit easier for a dev to grok?  But not both.

I guess given that cachedBech32Codec is not exported an options struct seem simpler?